File: 21_financial_technology/skills/02_trading_platforms/src/order_validator.java
// Enterprise order validation system
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public class EnterpriseOrderValidator {
    private static final int MIN_PRICE = 1;
    private static final int MAX_PRICE = 1000000;
    private static final int MIN_SIZE = 1;
    private static final int MAX_SIZE = 10000000;

    private final ConcurrentHashMap<String, Integer> positions;
    private final ConcurrentHashMap<String, Integer> notional;
    private final RiskLimits riskLimits;

    public EnterpriseOrderValidator(RiskLimits limits) {
        this.riskLimits = limits;
        this.positions = new ConcurrentHashMap<>();
        this.notional = new ConcurrentHashMap<>();
    }

    public ValidationResult validate(Order order) {
        ValidationResult result = new ValidationResult();

        // Price validation
        if (order.getPrice() < MIN_PRICE || order.getPrice() > MAX_PRICE) {
            result.addError("INVALID_PRICE");
        }

        // Size validation
        if (order.getQuantity() < MIN_SIZE || 
            order.getQuantity() > MAX_SIZE) {
            result.addError("INVALID_SIZE");
        }

        // Position limit check
        String symbol = order.getSymbol();
        int currentPos = positions.getOrDefault(symbol, 0);
        int newPos = currentPos + (order.getSide() == Side.BUY ?
                                   order.getQuantity() :
                                   -order.getQuantity());

        int posLimit = riskLimits.getPositionLimit(symbol);
        if (Math.abs(newPos) > posLimit) {
            result.addError("POSITION_LIMIT");
        }

        // Notional check
        double orderNotional = (double)order.getQuantity() * order.getPrice();
        String account = order.getAccount();
        double currentNotional = notional.getOrDefault(account, 0);
        double newNotional = currentNotional + orderNotional;

        double notionalLimit = riskLimits.getNotionalLimit(account);
        if (newNotional > notionalLimit) {
            result.addError("NOTIONAL_LIMIT");
        }

        // Account status
        if (!riskLimits.isAccountActive(account)) {
            result.addError("ACCOUNT_INACTIVE");
        }

        return result;
    }

    public void updatePosition(Order order) {
        String symbol = order.getSymbol();
        int qty = order.getSide() == Side.BUY ?
                  order.getQuantity() :
                  -order.getQuantity();

        positions.compute(symbol,
            (k, v) -> (v == null ? 0 : v) + qty);
    }

    static class ValidationResult {
        private java.util.List<String> errors = new java.util.ArrayList<>();

        public void addError(String error) {
            errors.add(error);
        }

        public boolean isValid() {
            return errors.isEmpty();
        }

        public java.util.List<String> getErrors() {
            return errors;
        }
    }

    static class RiskLimits {
        public int getPositionLimit(String symbol) { return 50000; }
        public double getNotionalLimit(String account) { return 100_000_000; }
        public boolean isAccountActive(String account) { return true; }
    }

    enum Side { BUY, SELL }

    interface Order {
        String getSymbol();
        String getAccount();
        int getQuantity();
        double getPrice();
        Side getSide();
    }
}

File: 21_financial_technology/skills/02_trading_platforms/src/order_validator.cpp
// Order validation engine
#include <atomic>
#include <map>

class OrderValidator {
private:
    static constexpr double MAX_PRICE = 10000.0;
    static constexpr int MAX_QUANTITY = 1000000;
    static constexpr double POSITION_LIMIT = 50000.0;

    std::map<std::string, std::atomic<int>> positions;

public:
    struct ValidationResult {
        bool valid;
        const char* reason;
    };

    ValidationResult validate_order(const char* symbol,
                                   int side,
                                   uint32_t quantity,
                                   double price) {
        // 1. Price check
        if (price <= 0 || price > MAX_PRICE) {
            return {false, "PRICE_OUT_OF_RANGE"};
        }

        // 2. Quantity check
        if (quantity <= 0 || quantity > MAX_QUANTITY) {
            return {false, "QUANTITY_OUT_OF_RANGE"};
        }

        // 3. Position limit check
        int current_position = positions[symbol].load();
        int new_position = current_position + (side == 1 ? quantity : -quantity);

        if (abs(new_position) > POSITION_LIMIT) {
            return {false, "POSITION_LIMIT_EXCEEDED"};
        }

        // 4. Duplicate check
        if (is_duplicate_order(symbol, quantity, price)) {
            return {false, "DUPLICATE_ORDER"};
        }

        return {true, nullptr};
    }

    void update_position(const char* symbol, int side, uint32_t quantity) {
        int qty_signed = (side == 1) ? quantity : -quantity;
        positions[symbol].fetch_add(qty_signed);
    }

private:
    bool is_duplicate_order(const char* symbol,
                           uint32_t quantity,
                           double price) {
        // Check recent order history for duplicates
        // Return true if duplicate found
        return false;
    }
};
